Proposed OLM-based Monitoring Stack Solution for RH Managed Services and future needs. (#866)
openshift/enhancements: enhancements/monitoring/monitoring-stack-operator.md

---
title: Monitoring for Red Hat Managed Services
authors:
- @bwplotka
- @fpetkovski
reviewers:
- @alanmoran
- @smarterclayton
- @kbsingh
- @pb82
- @simonpasquier
approvers:
- TBD
creation-date: 2021.08.12
last-updated: 2021.08.12
status: implementable
---

# Monitoring for Red Hat Managed Services

## Release Signoff Checklist

- [X] Enhancement is `implementable`
- [X] Design details are appropriately documented from clear requirements
- [X] Test plan is defined
- [ ] Operational readiness criteria is defined
- [ ] Graduation criteria for dev preview, tech preview, GA

> - [X] User-facing documentation is created in [openshift-docs]([https://github.com/openshift/openshift-docs/](https://github.com/openshift/openshift-docs/)) - (there are non required)

Reviewer: Where will the SRE/Managed Service owner-facing documentation be added?

## Summary

> This document proposes a solution and a delivery plan for scraping, querying and alerting on metrics related to Red Hat Managed Services running in OSD clusters. The proposed solution is not intended to complement the existing monitoring stack managed by the Cluster Monitoring Operator, nor used directly by Customers at this moment.

Reviewer 2: In an addon perspective, which SRE persona would likely consume the alerting part? For example, platform SREs, addon SREs or app SREs etc? The reason for asking is that Addons may have their own monitoring stack.

### Glossary

* MS-SRE/MT-SRE: Managed Service / Managed Tenant Site Reliability Engineer
* MTO: Managed Tenant Owners - the teams that develop each Managed Service. The Kafka Team is one example of an MTO.
* RHOBS: Red Hat Observability Service

> * RHODF: Red Hat Openshift Data Foundation

Reviewer 3: ```suggestion
* RHODF: Red Hat OpenShift Data Foundation
```

For additional acronyms, see [Red Hat Dictionary](https://source.redhat.com/groups/public/red-hat-dictionary-or-lexicon/red_hat_abbreviations_acronyms_and_initialisms_dictionary_wiki)

## Background

> Openshift Monitoring is currently composed of two main systems - Platform Monitoring (PM) and User Workload Monitoring (UWM), both of which are part of OpenShift core.

Reviewer: ```suggestion
OpenShift Monitoring is currently composed of two main systems - Platform Monitoring (PM) and User Workload Monitoring (UWM), both of which are part of OpenShift core.
```

Platform Monitoring holds metrics about core system components running in `openshift-*` and `kube-*` namespaces. Platform Monitoring is enabled in all OpenShift installations and cannot be disabled. Some of the components scraped by PM include core-dns, etcd, kube-apiserver, image-registry, operator-lifecycle-manager etc.
Platform Monitoring is by design an immutable stack, [which means that admins cannot extend it with additional metrics without losing the support.](https://docs.openshift.com/container-platform/4.7/monitoring/configuring-the-monitoring-stack.html#support-considerations_configuring-the-monitoring-stack)

User Workload Monitoring is an optional stack managed by Cluster Monitoring Operator (CMO) and is disabled by default. Once enabled by cluster admins, it allows users to monitor namespaces that are not part of OpenShift core.

> Please refer to the User Workload Monitoring [enhancement proposal](https://github.com/openshift/enhancements/blob/master/enhancements/monitoring/user-workload-monitoring.md) for more details. If enabled, only one stack can be deployed. The recommended way to get “platform” data is through Thanos Querier. Alternatively, Platform Monitoring Prometheus can also federate the latest data.

Reviewer 4: Isn't it simply reference to possibility to federate data from Prometheus instance(s) living in Platform Monitoring stack? E.g. if I'm not mistaken that's what Telemeter client uses.

> Please refer to the User Workload Monitoring [enhancement proposal](https://github.com/openshift/enhancements/blob/master/enhancements/monitoring/user-workload-monitoring.md) for more details. If enabled, only one stack can be deployed. The recommended way to get “platform” data is through Thanos Querier. Alternatively, Platform Monitoring Prometheus can also federate the latest data.

Reviewer 5: Is it officially supported? AFAIK, the federation endpoint of PM Prometheus is exposed only for telemeter usage.

> Please refer to the User Workload Monitoring [enhancement proposal](https://github.com/openshift/enhancements/blob/master/enhancements/monitoring/user-workload-monitoring.md) for more details. If enabled, only one stack can be deployed. The recommended way to get “platform” data is through Thanos Querier. Alternatively, Platform Monitoring Prometheus can also federate the latest data.

Reviewer 2: FYI the RHOAM addon (and since RHMI) which has GA'd and had been operating in production for a while now is using platform prometheus federation to gather platform metrics and complement with middleware metrics for alerting.

> Please refer to the User Workload Monitoring [enhancement proposal](https://github.com/openshift/enhancements/blob/master/enhancements/monitoring/user-workload-monitoring.md) for more details. If enabled, only one stack can be deployed. The recommended way to get “platform” data is through Thanos Querier. Alternatively, Platform Monitoring Prometheus can also federate the latest data.

Reviewer 3: The RHODS addon is doing the same.

Recently an essential model of managing software for customers emerged at Red Hat, referred to as “Managed Services” or “Managed Tenants”. From an architectural point there are three types of such services:

* Addons/Layered Services
* SaaS
* A combination of the first two (SaaS that requires an Addon)

The way these managed services are intended to run is presented in the following diagram.

> ![drawing](https://docs.google.com/drawings/d/e/2PACX-1vQrMBfwE8DJm3kVtC6SLVYJLEztJRABIYwvILMgtOmyyE6WqKSCqvZHksUsVUoqSZIt5E_4ro4ObGmY/pub?w=1312&h=511)

Reviewer 6: Please copy external references like diagrams into the repo with the design document, so they can be versioned and published without requiring access to Google Apps. Since there are quite a few figures for this enhancement, I suggest creating a folder with the same name as the base document and placing all of the images inside that folder.

We expect many more Addons to come in the following years.

## Motivation

Red Hat managed services are deployed in OSD clusters and run alongside user workloads.

Even though there are two monitoring stacks available in OSD clusters, neither is suitable for managed services. In OSD, Platform Monitoring is exclusively owned and managed by Platform SRE. It is customized to their needs and no modifications are allowed.
Because of this, managed tenant owners (MTOs) cannot create additional ServiceMonitors, Prometheus rules and other monitoring resources on metrics in the Platform Monitoring domain.

> User Workload Monitoring was designed to be primarily used by customers for monitoring their own workloads. Managed services are discouraged from using this stack since they could easily get DOS-ed by high cardinality data from end-users.

Reviewer 7: Actually user workload monitoring will interfere with the middleware monitoring stack employed by add-ons today and has to be disabled in order for the managed service to be installed.

Reviewer 7: I could be remembering that in reverse though.

Reviewer 2: To add here, UWM doesn't just scrape customer workload metrics. Technically, it scrapes every ServiceMonitor it can find outside the platform namespaces, assuming there were no feature updates. This has been an issue in RHOAM where it exposed the Addon's monitoring to the customers which was only intended to be used by Addon/RHOAM SREs.

Reviewer 8: Yes, this is correct and the we've already taken it into account independently of this enhancement. In 4.9 a namespace can be excluded from UWM by adding a label to it, which should solve the problem for addons.

This imposes an extra cost on teams developing and maintaining a managed service and increases time-to-market for new services.

### Goals

This enhancement aims to relieve the burden from managed service owners by providing a ready-made monitoring solution. Such a solution would encapsulate existing patterns already present in Platform Monitoring and User Workload monitoring.
Furthermore, it will reduce the operational complexity of running a managed service since the Managed Tenant SREs will not need to familiarise themselves with a new monitoring stack for each managed service.
In future it can be used to allow customers to define their monitoring stacks. UWM could use it in future underneath too. In details:

* Provide a way for managed services to scrape metrics from their own services and from platform components into prometheus-compatible storage.
* Allow managed service owners to query metrics from their services and platform components from prometheus-compatible storage.
* Provide a way for managed services to define alerting and recording rules on their own metrics and a subset of platform metrics.

> * Provide a way for managed services to create dashboards for monitoring their services.

Reviewer 9: Are there plans to provide OCP console integration? Currently the console queries thanos in cluster that provides both openshift-monitoring + openshift-user-workload-monitoring alerts for the console to display. I am unsure if this is desirable, just raising the point.

> * Allow aggregating portions of managed service metrics at a multi-cluster level, e.g. for reporting SLOs.

Reviewer 10: Not following this; what part of this effort will cover multi-cluster?

* Allow longer retention for portions of data (e.g. calculating 1 month burn down SLO budget).
* Allow analytics on portion of managed service data in conjuction with Telemeter data

> * End customers using UWM should not see managed service metrics.

Reviewer 11: Is this the same as "end customers using a managed service on their cluster that their workload talks to can't see the metrics of the managed service control plane?"  ultimately we want to ensure higher levels can aggregate workload metrics with the service metrics they consume (which may accidentally or deliberately be running on the same clusters as the workloads) - so just checking what this constraint means (i'm assuming the most minimal limit - "accidentally see")

Author: It is only about UWM, so user workload monitoring is assumed as things that can be used by customers and we don't want them to get confused by addons data.

Reviewer 12: Does this imply that managed services that want to expose users to metrics will need to do so through some other mechanism?

Author: I think it's orthogonal to what we discuss here with @smarterclayton  but good question too.

We answered it in `Non Goals`:

> * Expose managed service metrics to end-users. We know e.g. OCS addon wants this, but this can be achieved on top of the proposed solution using client or backend side filtering for data isolation if needed (e.g. prom-label-proxy). The monitoring team can do this in the next iteration.

* Soft Goal: A foundation for next iterations mentioned in ### Long term motivation.

### Long Term Motivation

To avoid maintaining multiple monitoring solutions, this proposal aims to be a foundation for future monitoring enhancements. Therefore, the following goals are left out of scope for the current enhancement but are kept in mind for future iterations:

* Such a solution as a community project which is easy to use & contribute to.

> * Reduce Platform Monitoring overhead and its single point of failure (SPOF) property.

Reviewer 9: From a workload perspective, will this run as an INFRA workload? or a WORKER workload?

Reviewer 2: +1 on this. Platform prom federation is not very reliable. RHOAM experienced/experiencing issues such as missing metric data at some period of time due to federation scraping timing issues or just the infra is under high load. This usually causes false-alerts in RHOAM.

### Non-Goals

* Access controls for managed service owners for metrics from different managed services. For example, there are currently no requirements for owners of managed services A to be prohibited from seeing metrics from managed service B, and vice versa.
* Expose managed service metrics to end-users. We know e.g. OCS addon wants this, but this can be achieved on top of the proposed solution using client or backend side filtering for data isolation if needed (e.g. prom-label-proxy). The monitoring team can do this in the next iteration.
* Integrate managed service metrics into the OpenShift console. We know OCS wants to integrate with OCP Console, but we assume the console team can help support switching between Prometheus-compatible data sources.
* Solve Platform Monitoring SPOF, scalability and overhead. This will be targeted in a follow-up enhancement.
* Allow logging and tracing capabilities. Other solutions should tackle this (designed by Logging and Tracing teams.)

## Proposal

### Personas

* OSD/Platform SRE
* MTSRE
* RHOBS Devs
* MTO
* Customer: User of Managed Service and OSD

### User Stories

#### Story 1

> As a MTSRE/MTO, I would like to define ServiceMonitors that will scrape metrics from my managed service by a Prometheus-compatible system.

Reviewer 2: I think that it will be helpful to specify the beginning of scraping and alerting as acceptance criteria here. When does a deployed monitoring stack start scraping and alerting?

The reason why I asked this is because, in the case of an addon installation, there is a possible window of unavailability of a monitoring stack. For example, RHOAM has its own monitoring stack deployed by the RHOAM operator, during this deployment process, RHOAM is mostly not observable if not observable at all because obviously the monitoring stack is being deployed. However, problems can arise in this stage and coupled with eventual updates from an addon's components, these issues can be very difficult to determine.

Will this be the same case for this proposed MonitoringStack or will it only guarantee its availability at some point or time T?

Reviewer 8: Great point. I see a potential overlap with meta-monitoring (monitoring of the monitoring stacks themselves) which we haven't fully explored yet. We intend to discuss this topic in tomorrow's working group meeting and we'll update the enhancement once we have an initial solution. Feel free to join if you're interested in contributing to the discussion :+1:

Reviewer 13: I want to challenge the interface we want to expose. Instead of `ServiceMonitor` it would be so much better to provide an `SLO`-based interface. The reasoning is that we want to abstract as much of the monitoring stack as we can. We want MTO to focus on SLOs and SLIs, not being Prometheus experts. The story I suggest is this one:

> As a MTSRE/MTO, I would like to define a SLO based CR that will create Prometheus-compatible config to scrape metrics from my managed service by a Prometheus-compatible system.

Example implementation:
- `PrometheusServiceLevel` CR from the sloth project: https://sloth.dev/examples/kubernetes/getting-started/

#### Story 3

As an MTSRE/MTO, I would like to create alerting and recording rules for Prometheus metrics exposed by my managed service.

#### Story 4

As an MTSRE, I would like to configure routing for alerts defined on Prometheus metrics exposed by my SaaS/Addon.

#### Story 5

As an MTO, I would like to query certain metrics for my service across all clusters in which the managed service is deployed or used.

#### Story 6

As an MTO, I would like to create dashboards for certain metrics across all clusters where the SaaS/Addon is deployed or used.

#### Story 7

As a Customer, I can use UWM without seeing Addons metrics, alerting or recording configurations.

### Implementation Details/Notes/Constraints

Since UWM and PM are part of Openshift Core, we would like to avoid increasing the complexity and introducing potential bugs for customers by extending those monitoring stacks. Because of this, we propose developing a new stack that will be able to support the managed service use-case while staying compatible with upcoming use-cases such as Hypershift or [KCP](https://github.com/kcp-dev/kcp).

#### Overview

> We propose developing a new open-source operator with a working name, “MonitoringStack Operator” (official name proposals are welcome!) that manages a `MonitoringStack` CRD and can be used to deploy an independent, prometheus-based stack for each managed service.

Reviewer 2: Can we possibly please clarify here for the case of more than 1 addon in a cluster? What does `stack` here means, is this referring to the set of ServiceMonitors per addon only or the prom/alertm instances?

The operator will be installed into its own namespace and will watch `MonitoringStack` resources in the entire cluster. Alongside itself, the operator will deploy two other operators:

* A Prometheus Operator in order to take advantage of the Prometheus management functionality already present in PO. An Alertmanager instance for managing alert routing will also be provisioned in the same namespace.
* A [Grafana Operator](https://operatorhub.io/operator/grafana-operator)

Once a `MonitoringStack` has been created, the operator will deploy the appropriate resources in the `MonitoringStackOperor`'s namespace based on how the stack is configured. In a simple case, this might be:

* a Prometheus CR
* a Grafana CR and a Grafana Data Source CR

> * a Thanos Querier if the stack is configured for high availability

Reviewer: Is there any drawbacks to making this opt-out instead of opt-in?

> In other cases, it might be just a Prometheus Agent deployment (forwarding only capabilities).

Reviewer 4: To me it is unclear which cases would deploying just this agent would take precedent over the former cases, would it be possible to extend this with an example or clarification?

Also is this already an existing solution or is this reference to this work in progress https://github.com/prometheus/prometheus/pull/8785? Maybe adding a note / link could be useful as well.

Reviewer 10: I read this as aligning this design with what we will need in HyperShift, where we don't want or need a full-blown monitoring stack on every cluster.  This way on a traditional topology, the MonitoringStack CR would state deploy the full-blown stack and on a hypershift topology, the MonitoringStack CR would deploy an agent that forwards off-cluster.

The `MonitoringStack` CRD is meant to be an abstraction over monitoring needs and requirements for a single tenant that specifies their needs. The underlying implementation is up to the MonitoringStackOperator.

> In this manner, managed service owners can simply create an instance of `MonitoringStack` alongside their services and delegate the provisioning and management of their monitoring stack to the operator.

Reviewer 2: A note here that besides a `managed service owner`,  a `managed service operator` or an `addon's operator` such as RHOAM (integreatly-operator) may need to determine/create monitoring stack configuration fields at operator runtime. For example, creating an Alertmanager email template with cluster-specific data.  So a MonitoringStack CR may need to be created dynamically.

Reviewer 8: This is a very good point. Our assumption is that each addon will be configured by an operator, so the operator will be able to create and update a MonitoringStack CR dynamically just like it can create and update any other kubernetes resource.

An example `MonitoringStack` resource could look as follows:

```yaml
---
apiVersion: monitoring.coreos.io/v1alpha1
kind: MonitoringStack
metadata:
    name: monitoring
    namespace: redhat-kafka
spec:
    retention: 12d
    resources:
      requests: 
        memory: 512M
      limits:  
        memory: 1G
---
```

> In the first iteration, each monitoring stack will be deployed and scoped to a single namespace. This means that the Prometheus instance will only be able to scrape monitoring resources from the namespace it is deployed in. In subsequent iterations, the CRD can be extended to allow scraping from multiple namespaces, either by namespace name or by namespace label selector.

Reviewer 3: For RHODS (Red Hat OpenShift Data Science), being able to scrape from multiple namespaces is critical.
The structure looks similar to the following.
1. redhat-ods-operator: This is where the RHODS Operator is deployed
2. redhat-ods-applications: This is where user-facing components and routes are deployed. User admins (via the `dedicated-admins` group) have access to some destructive privileges on this namespace as well.
3. redhat-ods-monitoring: This is where we currently deploy the monitoring stack. The reason we placed it in a separate namespace is so that end-users can't break the monitoring stack.

Reviewer 14: @philbrookes I believe this would be an issue for RHOAM too?

Reviewer 15: @alanmoran correct, RHOAM also scrapes from multiple namespaces

> In the first iteration, each monitoring stack will be deployed and scoped to a single namespace. This means that the Prometheus instance will only be able to scrape monitoring resources from the namespace it is deployed in. In subsequent iterations, the CRD can be extended to allow scraping from multiple namespaces, either by namespace name or by namespace label selector.

Reviewer 3: Writing the above also made me realise that we our current alerting depends on federation of metrics from cluster-monitoring, especially using kube-state-metrics. How do we address that need?

Reviewer 3: Ah, I see this need was addressed under `Scraping metrics from platform components`

With such a design, User Workload Monitoring could become just another instance of a `MonitoringStack`. Furthermore, Platform Monitoring can be trimmed down in the same way by allowing various OpenShift core components to deploy their `MonitoringStack` resource.

The following diagram illustrates a high-level overview of the final solution applied in the context of managed services.

![drawing](https://docs.google.com/drawings/d/e/2PACX-1vT4p10FDLEg0bv_pLUlwyVszq1275MNRJ-sBZMLxT5gcZ0q-tY8MHoEYpOlVmJuxlT40lqUQYCDwaSC/pub?w=1213&h=898)

> 

Reviewer 10: Nit:  It doesn't include the agent-based option.

_First iteration of the new monitoring operator, showing a monitoring stack with a Prometheus, Grafana and a shared Alertmanager._

We recommend using either [OperatorSDK](https://github.com/operator-framework/operator-sdk) or [KubeBuilder](https://github.com/kubernetes-sigs/kubebuilder) to develop the operator to increase development efficiency by taking advantage of the higher abstraction constructs, the rich documentation, and a vibrant community.

> Alternatively, we could also consider using a higher-level operator, such as [locutus](https://github.com/brancz/locutus), for rapid development with a path for GitOps deployments. In any case, this is an implementation detail that should be left to the team implementing this to decide on.

Reviewer: Nit: I'd love to see us being consistent with other Addons/RH-operators here if possible. 

New bundle/delivery strategies sound like a small change but require training a lot of people across several orgs.

## Design Details

### Installation

The newly developed operator will be deployed using the Operator Lifecycle Manager (OLM).

This will allow us to make releases much faster and iteratively and incrementally. As a result, we will be able to respond to new requirements much more rapidly. Release batches will also become smaller, reducing the risk of multiple independent features affecting each other.

> In order to be aligned with the descoping [plan](https://docs.google.com/presentation/d/1j1J575SxS8LtL_YvKqrexUhso7j4SgrLfyNrDUroJcc/edit#slide=id.ga089527607_0_0) that the OLM team is currently working towards, we recommend avoiding `OperatorGroups` as they are going to become deprecated in the future (possibly as soon as OpenShift 4.11).

Reviewer 16: `OperatorGroups` is unclear to me. Does this need a definition?

For this reason, the newly developed operator should be installed globally in a cluster and should provide its own CRDs and Roles as part of the installation bundle. A cluster administrator would then create RoleBindings to bind the roles to namespaces to which the operator needs access.
More details about the descoping plans and strategies are currently described can be found in this [hackmd](https://hackmd.io/wVfLKpxtSN-P0n07Kx4J8Q?view#Descoping-Plan) document and are also illustrated in a short [slide-deck](https://docs.google.com/presentation/d/1j1J575SxS8LtL_YvKqrexUhso7j4SgrLfyNrDUroJcc/edit#slide=id.ga089527607_0_0).

### Metrics retention

> Since each addon can create its own monitoring stack, MTOs will be able to configure retention for their service as they see fit.

Reviewer: Would it be worth specifying a default here? Default to `30d` and then the MTO can override if needed.

### Global view

> In order for MTSRE/MTO to be able to view portions of metrics for all of their deployments, each managed service will have its own tenant in Red Hat Observability (RHOBS). This will allow remote-writing a subset of metrics from the Prometheus instances into RHOBS.

Reviewer: Do we plan to track the changes needed into RHOBS separately? Is there a design doc for how this tenancy would look like?

Reviewer 4: I guess [Observatorium](https://github.com/observatorium/observatorium) could be a more apt reference, since Observatorium (more precisely the [API](https://github.com/observatorium/api)) is what enables these features already at the present (although I'm unable to find any further design doc at this moment, regarding specifically the tenancy feature). RHOBS is then our hosted instance of Observatorium. 

Maybe expanding on this / adding few words / adding references on RHOBS <-> Observatorium here or elsewhere in the document could be beneficial.

Reviewer 4: nit 
```suggestion
In order for MTSRE/MTO to be able to view portions of metrics for all of their deployments, each managed service will have its own tenant in Red Hat Observability Service (RHOBS). This will allow remote-writing a subset of metrics from the Prometheus instances into RHOBS.
```

> In order for MTSRE/MTO to be able to view portions of metrics for all of their deployments, each managed service will have its own tenant in Red Hat Observability (RHOBS). This will allow remote-writing a subset of metrics from the Prometheus instances into RHOBS.

Reviewer 10: I know I've seen this discussed on slack but is this decision captured somewhere as part of a managed services onboarding flow, along with any budget numbers a new service team would be responsible for covering?

### Scraping metrics from platform components

> MTOs often need to create alerts and dashboards which include metrics coming from platform components. A typical example is using CPU and memory resource metrics from kube-state-metrics. For ease of use, MTOs will be able to define allow-listed scrape on Platform Monitoring /federate endpoint to access relevant metrics.

Reviewer: > MTOs will be able to define allow-listed scrape on Platform Monitoring /federate endpoint
Where would such an allow-list be defined? Who controls it?

Do we have any concerns if there's half a dozen Addons running on the cluster, each with their own `MonitoringStack`, all federating a bunch of platform metrics? What's the contract between managed services and the platform in this case, since `/federate` queries can be pretty resource-intensive?

Reviewer 2: One thing to consider here is also the `/federate` scrape timing between platform prom and addon prom. IOW, how fast can `/federate`  respond to the addon's scrape timing. For multiple addons using `/federate`, this could become an SPOF problem or as different addons come to existence, the timing would dynamically change.

> In the first iteration, this can be implemented by defining a ServiceMonitor against the `/federate` endpoint of the Platform Monitoring Prometheus. Improvement on this side will be revisited in the next enhancement (e.g ability to scrape those data directly from sources).

Reviewer 17: The second step might be challenging since platform targets require tokens or client TLS certificates (and eventually everything should use TLS).

Author: Let's explain challenges we might have (non cached scrape endpoints e.g kubelet, node-exporter), but we envison to tackle it later 👍🏽

> ### High Availability

Reviewer: How will scalability be handled? Is there anything we can do if the disks for the Prometheus instances fill up, etc? 

If we're going for one `MonitoringStack` per Addon, SRE would then end up having to care about `Customers * number of Addons * replicas` Prometheis (!)

All components except Prometheus are stateless and can run with more than one replica. In order to make Prometheus highly-available, a similar approach to the one in CMO will be used. The monitoring operator will provide a Thanos Querier for each `MonitoringStack` and a sidecar next to each Prometheus.
A Thanos Ruler is not needed in this architecture since all recording rules and alerts will be evaluated locally in each Prometheus instance.

### Alert Routing

> For simplicity reasons, routing alerts that come from managed service monitoring stacks will be handled by a single Alertmanager.

Reviewer: What namespace would this alertmanager live in?

We can take advantage of the recently added <code>[AlertmanagerConfig](https://github.com/prometheus-operator/prometheus-operator/blob/master/Documentation/api.md#alertmanagerconfig)</code> and allow MTO/MTSRE to create their own routing configuration in a centrally deployed Alertmanager instance.

> Finally, the MTSRE team could centrally configure receivers in one place for all managed services and service owners would simply use those receivers in their routing rules. It is worth noting that the last feature depends on [https://github.com/prometheus-operator/prometheus-operator/issues/4033](https://github.com/prometheus-operator/prometheus-operator/issues/4033).

Reviewer 8: @simonpasquier is it possible to use both the secret _and_ an `AlertmanagerConfig`? I thought that with the `AlertmanagerConfig` everything gets scoped to the namespace from which the config is coming. If that is the case, how would a tenant reference a receiver from the common secret in their `AlertmanagerConfig`?

> In addition to the built-in Alertmanager from MSO, users will have the option to route alerts to external Alertmanagers, such as the PM Alertmanager.

Reviewer: I haven't seen this use case so far either. Is there a concrete use case we can point to?

#### Notable Alerting Routing Alternative

During the Monitoring Enablement Working Group meetings, we received feedback from both Platform SRE and Managed Tenant SRE that Alertmanager, in general, has a low footprint, and the Prometheus Operator does a good job of upgrading it and keeping it running.

Moreover, Platform SRE expressed concerns that a shared Alertmanager configuration could lead to conflicts or misconfiguration that can impact alerting for every managed tenant running in the cluster.

> Based on the information, we do not see strong reasons not to have one Alertmanager per managed service. Either approach should work as an initial solution.

Reviewer 2: Maybe a little bit inconsistent with line 243?

Author: right, clarified. Thanks (:

### Creating Dashboards

The OCP Console does not (yet) support querying different Prometheus instances, making it hard to use it for creating dashboards.

> For this reason, the monitoring operator will optionally deploy a Grafana operator that can manage Grafana instances in the namespaces where a MonitoringStack CR is present. Managed service SREs and owners will then be able to use the CRDs provided by the Grafana operator to create their own dashboards.

Reviewer: One Grafana instance per Addon? Again, need to think about the UX a bit here. Backplane already makes it hard enough to get into a single grafana.

Having to hop across Grafana instances during an incident could lead to some human errors and frustration.

Would it be possible to have a single Grafana and the Addons simply add their own datasource to this Grafana? 

This would allow for some use cases like local cluster-state dashboards that span several Addons.

Reviewer 10: I thought we were removing grafana from OCP console ... now we are adding it back (albeit like a side car?) who owns this grafana operator for RHT?  Who supports it?

Author: We don't add it back to Core. It's similar support as MonitoringStack. We will support it together with other existing maintainers.

Author: Open question: 
* Do we then should descope Grafana (and similarly AM) from MonitoringStack CR and ask users to create AM CR (from PO) and Grafana instances (from Grafana Opetator) directly? It feels more natural and composable TBH.

Monitoring Stack can still create data sources CRs for Grafana integration.

Reviewer 8: @aditya-konarde I agree that one Grafana per addon might not be ideal for MT-SRE. However, we would like to look at a `MonitoringStack` resource as independent of the Managed Services use case and even decoupled from OpenShift. In the future, we would like to be able to reuse this abstraction for different use cases.

If we allow end-users to deploy `MonitoringStack`s, their dashboards will automatically end up in this central Grafana instance. As an alternative, can MT-SRE simply deploy a new Grafana instance through the Grafana Operator and select datasources from all Addon namespaces. This would keep the `MonitoringStack` resource composable and reusable instead of coupling it to a single use case.

Author: I like this, but the problem I see is that MonitoringStack now is just a small wrapper/abstraction over Prometheus 🤔

Author: Which technically could be just a next PO CRD

In addition to this, several teams in the OpenShift organization already deliver their software through OLM, some of which are also in the observability realm. Specifically, the logging team has been delivering their stack in this fashion for a while, making them a good example.

#### Feature duplication with CMO

The Cluster Monitoring Operator has grown in features over time and the codebase contains functionality that might also be needed for the Monitoring Stack Operator. Over time, we could end up implementing the same features in two different projects.

One solution which we can aim for long-term would be to deploy User Workload Monitoring as another `MonitoringStack` instance. CMO and users can still deploy UWM will be able to configure it the same way as today, only CMO would delegate the provisioning and management to MSO by creating a `MonitoringStack` resource.

#### Resource usage overhead: Prometheus

> Deploying a single prometheus instance for each managed service does come with certain resource overhead. Even though the data will be perfectly sharded, there is still a certain fixed resource requirement for simply running Prometheus. In addition to this, sharing resources is usually more efficient because it reduces idle time.

Reviewer 16: ```suggestion
Deploying a single prometheus instance for each managed service does come with certain resource overhead. Even though there will be no unintentional data duplication, the additional processes running prometheus and other components will increase the per-metric-ingested resource overhead.
```

> In order to evaluate the resource overhead that comes with functional sharding, we did an experiment where we deployed one Prometheus instance scraping metrics from 3 exporters (node_exporter, kube-state-metrics and kubelet) alongside 3 independent Prometheus instances, each scraping data from only one exporter.

Reviewer 16: ```suggestion
In order to evaluate this, we deployed one Prometheus instance scraping metrics from 3 exporters (node_exporter, kube-state-metrics and kubelet) alongside 3 independent Prometheus instances, each scraping data from only one exporter.
```

Reviewer 16: I think it would also help to name the four prometheus instances according to the drawing below.

We then visualized the memory usage of each Prometheus instance over time and the results are shown on the image below.

![alt_text](https://docs.google.com/drawings/d/e/2PACX-1vTQR13HcQ_wbAqVdX7rRAGBXxRJmrhN-DBMLQ1i92DqjIFVUD05uIsezacTilkZM7CwMP6XHVh4u3Uo/pub?w=1440&h=813)

> At the latest timestamp, the total memory usage for the 3 instances forming functional sharding is 434MB, while the memory usage of the single instance is 316MB. What is also worth pointing out, and is already illustrated by the graph, is that systems that do not export many metrics, such as node_exporter and kubelet, require small Prometheus instances in terms of memory requirements.

Reviewer 16: > What is also worth pointing out, and is already illustrated by the graph, is that systems that do not export many metrics, such as node_exporter and kubelet, require small Prometheus instances in terms of memory requirements.

I guess the advantage we get through this is more scheduling options with a wider variety of machine types?

> We also performed some back-of-the envelope calculations in order to illustrate the added cost of functional sharding in terms of actual money. A `c5.xlarge` on-demand instance in AWS (a compute optimized instance with 4 CPUs and 8GB RAM) costs 140$ per month or 18$ per GB per month.

Reviewer 4: Nit 
```suggestion
We also performed some back-of-the-envelope calculations in order to illustrate the added cost of functional sharding in terms of actual money. A `c5.xlarge` on-demand instance in AWS (a compute optimized instance with 4 CPUs and 8GB RAM) costs 140$ per month or 18$ per GB per month.
```

When using reserved capacity, the cost goes down to 11$ per GB per month for the same instance with a 1 year reservation.

Our conclusion based on these factors is as follows: while sharing a Prometheus instance does come with better resource utilization, we do not believe it justifies the reduced resiliency that it also brings along.

> The resiliency of such a shared instance would be further reduced as more managed services are installed in a cluster. In addition, a shared Prometheus instance is often more volatile regarding resource utilization, which leads to users over provisioning resources to prevent OOMs by misconfiguration or complex querying patterns.

Reviewer 10: I think another key upside is the separation of concerns, allowing autonomy, control and responsibility to be split among the different consuming personas.

Reviewer 8: Good point, added the suggestion.

> Finally, even if at one point functional sharding does end up being a problem from a resource usage perspective, the monitoring operator will nevertheless support deploying multi-namespace monitoring stacks. This will effectively support the pattern of sharing a prometheus instance between managed services.

Reviewer 16: I think another advantage worth pointing out is that we get much finer grained control over how much storage is used for metric retention. With the proposed system we can control on a per ServiceMonitor level how long the respective metrics should be kept.

Reviewer 16: ```suggestion
Finally, with the proposed stack we retain the option of deploying multi-namespace monitoring stacks. This will effectively support the pattern of sharing a prometheus instance between managed services.
```

### Migration

There are multiple monitoring solutions that Addons currently use:

* UWM and AppSRE has their monitoring stack for SaaS services, which they plan to migrate to UWM

>   * MSO is planned to have feature parity with UWM with the exception of global view with Platform Monitoring. We will initially use federation to close this gap (See )

Reviewer 10: Just for context; SD were told in late 2019 that UWM would be made to fit our use-cases, and it never was.  There needs to be some lessons taken away from this, particularly on the prioritization side.  In the end, there was a lot of wheel-spinning, frustration and effort spent in the intervening months and years that we can't afford to repeat.

>   * MSO is planned to have feature parity with UWM with the exception of global view with Platform Monitoring. We will initially use federation to close this gap (See )

Author: 🔥 

Great point, fully agree (:

* Platform Monitoring
  * The PM stack cannot be used for user/managed service monitoring anyway. Therefore no migration is predicted so far.
* Observability Operator
  * This is great work done by the Kafka team to get unblocked and as we see it has feature parity to UWM with addition of Dashboarding and Promtail (logging aggregation). MSO will provide all but logging integration. For the logging part, Openshift Logging Operator should be leveraged.

For other cases we are not aware of, we will be providing a weekly [Working Group](https://docs.google.com/document/d/1fVDDn9b69mXNyfIFpuCoOHasY0pGkDyuJ6qUyPySSdI/edit#) for internal teams as well as for building a community to support it.

In all cases, MSO should not conflict with any other monitoring tool installed in the cluster and each managed service can adopt the monitoring stack on their own time. The monitoring team will support them in the migration process.

### Documentation Required

* Project documentation on GitHub + website
* Tutorial on how to run it on Openshift
* Onboarding RHOBS documentation

### Open Questions

* If customer can use both MSO and UWM in the future, which one should they use?

> * Should we allow customer to create MSO CRs? (MonitoringStacks).

Reviewer: Flip question: How do we prevent them from doing that :)

Reviewer 8: If we decide to deploy a central Grafana instance, it would be a bit weird to allow end users to deploy a monitoring stack CR. Suddenly MTSREs would see datasources from end user stacks, while end-users will not able to create dashboards for their own Prometheus instances.

Author: Yes. That's why I think we are leaning into the idea of enabling users to define their own, multi-ns/shared Grafana instances, outside of Monitoringstack TBH. WDYT?

* How will it fit into Hypershift? Will it have enough flexibility to support the Hypershift use case?
  * Full Hypershift proposal is in progress, but the whole stack can be simply run on DataPlane with all capabilities except kube-state-metrics. Given that it is unknown how Addons will run their services on Hypershift, discussing this would be premature.

### FAQ

Waiting for questions first (:

### Test Plan

The proposed solution will be tested with two suites:

* Unit tests - running independently of each other and without side effects
* End to end tests - in the same repository as the managed services monitoring operator. These tests will be executed against an actual kubernetes cluster and are intended to verify end-to-end functionality of the entire system.

## Implementation History

We plan to develop the solution in multiple deliverable milestones:

*No Alert Routing, No HA, No Dashboarding:*

1. Create GitHub Project, CI setup
2. Create OLM bundle with Monitoring Stack operator and basic CR
3. Allow OLM bundle to deploy or require further operators (Grafana, Prometheus)

> 4. MonitoringStack CR constructing simple Prometheus stack without HA

Reviewer 18: Currently Observatorium MST tenant expects the clients (for remoteWriting) to know the service accounts that are created by SSO. For OSD metrics, the platform monitoring stack explicitly injects these service accounts for remortWriting. With monitoring-stack OLM operator, its going to be tricky to bundle these sensitive service accounts as part of the OLM bundle.

> 4. MonitoringStack CR constructing simple Prometheus stack without HA

Author: Yup, we are not going to add any secrets to OLM bundle. We plan to support bundle to read secrets from e.g Kubernetes secret and inject with Prometheus (OIDC/token). (:

* With single (or HA) Prometheus, this stack will not scale. In order to scale we would need to deploy Observatorium which is [explained](#full-forwarding-mode-with-local-external-or-hosted-observatorium) in a previous alternative.
* The same topology will be reproducible with the proposed alternative by specifying multiple namespaces in the `MonitoringStack` resource.

### Shared Thanos Querier

Pros:

* Less resource overhead
* Global view out of the box

Cons:

* This approach makes a `MonitoringStack` coupled to a centralized component. In order for individual teams to query their own metrics, they will need to query all prometheus instances.
* Currently Thanos Query does not have good QoS mechanisms for independent query targets. Intensive queries from one team could therefore impact querying capabilities of other teams.
* Since we want to build a community around the project, this approach makes the stack overall less composable and less reusable. It assumes that `MonitoringStack` owners are fine with exposing their metrics to each other. This might be the case for Managed Services, but it does not necessarily apply in the general sense.

> * In case a central view is needed, an additional Thanos Query instance can still be deployed independently.

Reviewer: I think this will be a much better UX from the SRE perspective, although I understand the resource/RBAC constraints. Is this something we'd like to add to the long term considerations? 

I can imagine a good way to deploy such a querier would be through the operator, so that any new `MonitoringStack` Prometheis are added automatically as targets to the querier.

Reviewer 8: Do we see a future where not all addons will be managed by the same SRE team? IOW, could we have MTSRE-1 responsible for addons 1, 2 and 3, and MTSRE-2 responsible for addons 4, 5, and 6?

Author: I think anyone can add Thanos Querier on top on their own (we can add that docs,) plus we talked about shared Grafana instances that will allow you to at least combine multiple sources under single dashboard, so it should be good experience (:

### Deploy a third Prometheus instance through Cluster Monitoring Operator

While this is a viable alternative, deploying a third instance through CMO will merely be a stop-gap solution that can be challenging to reuse for upcoming use-cases, such as hypershift. In addition, coupling managed service monitoring to OCP releases will reduce delivery velocity, thereby inhibiting agile software development and iterative value delivery.
An OLM-based stack, on the other hand, allows us to release as often as possible and deliver value to managed service owners faster and in incremental batches.

Furthermore, CMO in the current state already takes on a lot of responsibility. It defines and manages two stacks, Platform Monitoring and User Workload Monitoring. Expanding the scope of CMO further would make it harder to maintain in the long run.

### Upgrade / Downgrade Strategy

New versions of the managed service monitoring stack will be published in OLM. Upgrades and downgrades will be performed by OLM itself.

### Graduation Criteria

TBD

#### Dev Preview -> Tech Preview

TBD

#### Tech Preview -> GA

TBD

#### Removing a deprecated feature

TBD

> ### Version Skew Strategy

Reviewer: How do we plan to address API version drifts between the built in CMO vs the MonitoringStack Operator? Does the MSO simply not bring any CRD's along?

Author: This has to be solved with OLM help. The new scoped/descoped semantics should help.

Author: Still, something to clarify

* Such a solution as a community project which is easy to use & contribute to.
* Reduce Platform Monitoring overhead and its single point of failure (SPOF) property.
* Auto-scaling metric scraping based on number of ServiceMonitors
* Soft multi-tenancy for local storage (TSDB).
* Opt-in for full forwarding mode.
* Full architecture for Hypershift ([WIP](https://docs.google.com/document/d/1BJarERppgiJ8esc6d8anbJMOQ0AflFBQea-Zc9wAp0s/edit))

### Non-Goals

* Access controls for managed service owners for metrics from different managed services. For example, there are currently no requirements for owners of managed services A to be prohibited from seeing metrics from managed service B, and vice versa.
* Expose managed service metrics to end-users. We know e.g. OCS addon wants this, but this can be achieved on top of the proposed solution using client or backend side filtering for data isolation if needed (e.g. prom-label-proxy). The monitoring team can do this in the next iteration.

> * Integrate managed service metrics into the OpenShift console. We know OCS wants to integrate with OCP Console, but we assume the console team can help support switching between Prometheus-compatible data sources.

Reviewer 19: > Do you mean Thanos Querier? If yes, then we assume the PM one? (Platform Monitoring)
Yes I meant "Querier" not "ruler"

> We believe there should be an effort for OpenShift Console to discover MonitorStacks that comes from addons that need extra widget/dashboard functionality on console based on that and use that endpoint instead of PM Thanos Querier.
My understanding is that this is the exact thing they object to as most of the time you will need some of the UI to work against one querier while some other part of the UI against another. And the console was not built to do that. 
For UWM they just use the same querier as they use for PM

My position is that if we will wait for them to do this kind of work it might ship far far into the future. I think this proposal needs to have a goal that says that this new monitoring stack needs to expose whatever is needed for remote writes to the Thanos instance used by the console. I think that most of the effort will go into autentication

Once a `MonitoringStack` custom resource (CR) has been created, the operator will deploy the appropriate resources in the `MonitoringStackOperor`'s namespace based on how the stack is configured. In a simple case, this might be:

* a Prometheus CR
* a Grafana CR and a Grafana Data Source CR
* a Thanos Querier if the stack is configured for high availability

In other cases, it might be just a Prometheus Agent deployment (forwarding only capabilities).

The `MonitoringStack` CRD is meant to be an abstraction over monitoring needs and requirements for a single tenant that specifies their needs. The underlying implementation is up to the MonitoringStackOperator.

In this manner, managed service owners can simply create an instance of `MonitoringStack` alongside their services and delegate the provisioning and management of their monitoring stack to the operator.

An example `MonitoringStack` resource could look as follows:

```yaml
---
apiVersion: monitoring.coreos.io/v1alpha1
kind: MonitoringStack
metadata:
    name: monitoring
    namespace: redhat-kafka
spec:
    retention: 12d
    resources:
      requests: 
        memory: 512M
      limits:  
        memory: 1G
---

> ```

Reviewer 19: ODF-MS needs support for label matching on monitoring resources. Without it, the monitoring stack will take into account every single monitoring resource (PodMonitor, ServiceMonitor, and PromRule) deployed in the namespace - This is not the desired behaviour. 

Our monitoring resources are deployed and (reconciled - so we cannot just delete them) by the underlying OCS product we are deploying. But from a managed service perspective not all of them are relevant and some of them might even create false-positive alerting (SRE incidents) and strain on the prom instance. 
The way we handle it today is by labelling (by our umbrella/addon operator) only the monitoring resources relevant for the managed service operation and instructing our monitoring stack to select only these (using selectors)

Losing this ability might create supportability issues for ODF-MS

Finally, with the proposed stack we nevertheless retain the option of deploying multi-namespace monitoring stacks. This will effectively support the pattern of sharing a Prometheus instance between managed services.

### Migration

There are multiple monitoring solutions that Addons currently use:

* AppSRE have their monitoring stack for SaaS services, which they plan to migrate to UWM
  * MSO is planned to have feature parity with UWM with the exception of global view with Platform Monitoring. We will initially use federation to close this gap (See )
* Platform Monitoring
  * The PM stack cannot be used for user/managed service monitoring anyway. Therefore no migration is predicted so far.
* Observability Operator
  * This is great work done by the Kafka team to get unblocked and as we see it has feature parity to UWM with addition of Dashboarding and Promtail (logging aggregation). MSO will provide all but logging integration. For the logging part, Openshift Logging Operator should be leveraged.

> 

Reviewer 19: I don't see a mention of the ODF-MS monitoring stack solution, which to me hints that we never verified feature parity with that solution. It is similar to the "Observability Operator" but does differ on some points

Author: I think we discussed your requirements e.g in ODF Managed Service SD sync, so I don't think you are forgotten. We addressed all your questions in this proposal too. *Did we miss anything?*

We don't have concrete clarity on your current ODF-MS monitoring stack solution. As far I as I remember it was based on PM (Platform Monitoring) which was primarily not designed for this work. Is there any extra functionality of ODF-MS that is worth mentioning here?

Reviewer 19: ODF-MS monitoring stack is not based on PM. I think the main thing that is missing is around "Monitoring Resource" selection and OCP console integration. I do believe I have left separate comments addressing these concerns.

Author: perect, thanks.

## Background

OpenShift Monitoring is currently composed of two main systems - Platform Monitoring (PM) and User Workload Monitoring (UWM), both of which are part of OpenShift core.

Platform Monitoring holds metrics about core system components running in `openshift-*` and `kube-*` namespaces. Platform Monitoring is enabled in all OpenShift installations and cannot be disabled. Some of the components scraped by PM include core-dns, etcd, kube-apiserver, image-registry, operator-lifecycle-manager etc.
Platform Monitoring is by design an immutable stack, [which means that admins cannot extend it with additional metrics without losing the support.](https://docs.openshift.com/container-platform/4.7/monitoring/configuring-the-monitoring-stack.html#support-considerations_configuring-the-monitoring-stack)

User Workload Monitoring is an optional stack managed by Cluster Monitoring Operator (CMO) and is disabled by default. Once enabled by cluster admins, it allows users to monitor namespaces that are not part of OpenShift core. If enabled, only one stack can be deployed.

> The recommended way to get “platform” data is through Thanos Querier. Alternatively, the UWM stack can scrape the `/federate` endpoint expsed by the Platform Monitoring Prometheus to injest a subset of the available platform metrics.

Reviewer 5: ```suggestion
The recommended way to get “platform” data is through Thanos Querier. Alternatively, the UWM stack can scrape the `/federate` endpoint exposed by the Platform Monitoring Prometheus to ingest a subset of the available platform metrics.
```

* Integrate managed service metrics into the OpenShift console. We know OCS wants to integrate with OCP Console, but we assume the console team can help support switching between Prometheus-compatible data sources.
* Solve Platform Monitoring SPOF, scalability and overhead. This will be targeted in a follow-up enhancement.
* Allow logging and tracing capabilities. Other solutions should tackle this (designed by Logging and Tracing teams.)

## Proposal

### Personas

* OSD/Platform SRE
* MTSRE
* RHOBS Devs
* MTO
* Customer: User of Managed Service and OSD

### User Stories

#### Story 1

As a MTSRE/MTO, I would like to define ServiceMonitors that will scrape metrics from my managed service by a Prometheus-compatible system.

#### Story 2

As an MTSRE/MTO, I would like to query SaaS/Addon metrics using PromQL to understand the health or troubleshoot the performance of my managed service.

#### Story 3

> As an MTSRE/MTO, I would like to create alerting and recording rules for Prometheus metrics exposed by my managed service.

Reviewer 13: Again proposing to abstract alerting and recording rules away, in favor of `PrometheusServiceLevel` CR or something similar (sloth):

> As an MTSRE/MTO, I would like to create SLO-oriented CRs, that will translate to Prometheus alerting and recording rules for my managed service.

#### Story 4

> As an MTSRE, I would like to configure routing for alerts defined on Prometheus metrics exposed by my SaaS/Addon.

Reviewer 13: Again proposing to abstract routing away, in favor of PrometheusServiceLevel CR or something similar (sloth):

> As an MTSRE/MTO, I would like to create SLO-oriented CRs, that will configure alerts defined on Prometheus metrics exposed by my SaaS/Addon.

#### Story 5

As an MTO, I would like to query certain metrics for my service across all clusters in which the managed service is deployed or used.

#### Story 6

> As an MTO, I would like to create dashboards for certain metrics across all clusters where the SaaS/Addon is deployed or used.

Reviewer 13: I think SLO dashboard can be abstracted as well. Creating grafana dashboards is SO time consuming, and as an MTSRE engineer, I want to have an homogeneous environment when switching from one addon to another. Plus, we do not want MTO to become grafana dashboard experts, or spend any time on this. This is why I would HIGHLY suggest:

> As an MTO, I would like to provide an SLO based CR, that automatically creates dashboards that allow me to visualize SLO/SLI and error budgets for certain metrics across all clusters where the SaaS/Addon is deployed or used.

**Example implementation:**
- Sloth SLO oriented grafana dashboards: https://sloth.dev/dashboards/

#### Story 7

As a Customer, I can use UWM without seeing Addons metrics, alerting or recording configurations.

> 

Reviewer 13: Adding an extra requirement for MTSRE/MTO:

```markdown
### Story 8

As an MTO, I would like to provide an SLO based CR, that is actively reconciled by a kubernetes-based operator, so that I can update my SLO/SLI configurations smoothly.
```

**Example implementation:**

- sloth operator/controller: https://sloth.dev/usage/kubernetes/

The following diagram illustrates a high-level overview of the final solution applied in the context of managed services.

![drawing](./assets/monitoring-stack-diagram.png)

_First iteration of the new monitoring operator, showing a monitoring stack with a Prometheus, Grafana and a shared Alertmanager._

## Design Details

### Installation

The newly developed operator will be deployed using the Operator Lifecycle Manager (OLM).

This will allow us to make releases much faster and iteratively and incrementally. As a result, we will be able to respond to new requirements much more rapidly. Release batches will also become smaller, reducing the risk of multiple independent features affecting each other.

In order to be aligned with the descoping [plan](https://docs.google.com/presentation/d/1j1J575SxS8LtL_YvKqrexUhso7j4SgrLfyNrDUroJcc/edit#slide=id.ga089527607_0_0) that the OLM team is currently working towards, we recommend avoiding [`OperatorGroups`](https://docs.openshift.com/container-platform/4.2/operators/understanding_olm/olm-understanding-operatorgroups.html).
There is a good chance that they will become deprecated in the future (possibly as soon as OpenShift 4.11).

> For this reason, the newly developed operator should be installed globally in a cluster and should provide its own CRDs and Roles as part of the installation bundle. A cluster administrator would then create RoleBindings to bind the roles to namespaces to which the operator needs access.

Reviewer 5: > For this reason, the newly developed operator should be installed globally

Isn't conflicting with the below statement?

> In the first iteration, each monitoring stack will be deployed and scoped to a single namespace.

Reviewer 5: > For this reason, the newly developed operator should be installed globally in a cluster and should provide its own CRDs

Does it mean the new operator has to ship Prometheus related CRDs(PromethesRule, AMConfig,..) ? If so, wouldn't cause conflict with CRDs which already shipped by CMO?

  See [Documentation Required](#documentation-required) for further details.

## Summary

This document proposes a solution and a delivery plan for scraping, querying and alerting on metrics related to Red Hat Managed Services running in OSD clusters. The proposed solution is not intended to complement the existing monitoring stack managed by the Cluster Monitoring Operator, nor used directly by Customers at this moment.

### Glossary

* MS-SRE/MT-SRE: Managed Service / Managed Tenant Site Reliability Engineer
* MTO: Managed Tenant Owners - the teams that develop each Managed Service. The Kafka Team is one example of an MTO.
* RHOBS: Red Hat Observability Service
* RHODF: Red Hat OpenShift Data Foundation

For additional acronyms, see [Red Hat Dictionary](https://source.redhat.com/groups/public/red-hat-dictionary-or-lexicon/red_hat_abbreviations_acronyms_and_initialisms_dictionary_wiki)

## Background

OpenShift Monitoring is currently composed of two main systems - Platform Monitoring (PM) and User Workload Monitoring (UWM), both of which are part of OpenShift core.

> Platform Monitoring holds metrics about core system components running in `openshift-*` and `kube-*` namespaces. Platform Monitoring is enabled in all OpenShift installations and cannot be disabled. Some of the components scraped by PM include core-dns, etcd, kube-apiserver, image-registry, operator-lifecycle-manager etc.

Reviewer 20: ```suggestion
Platform Monitoring holds metrics about core system components running in `openshift-*` and `kube-*` namespaces. Platform Monitoring is enabled in all OpenShift installations and cannot be disabled. Components scraped by PM include core-dns, etcd, kube-apiserver, image-registry, operator-lifecycle-manager, etc.
```

> Platform Monitoring is by design an immutable stack, [which means that admins cannot extend it with additional metrics without losing the support.](https://docs.openshift.com/container-platform/4.7/monitoring/configuring-the-monitoring-stack.html#support-considerations_configuring-the-monitoring-stack)

Reviewer 20: ```suggestion
Platform Monitoring is by design an immutable stack, [which means that admins cannot extend it with additional metrics without losing support.](https://docs.openshift.com/container-platform/4.7/monitoring/configuring-the-monitoring-stack.html#support-considerations_configuring-the-monitoring-stack)
```

User Workload Monitoring is an optional stack managed by Cluster Monitoring Operator (CMO) and is disabled by default. Once enabled by cluster admins, it allows users to monitor namespaces that are not part of OpenShift core.

> Please refer to the User Workload Monitoring [enhancement proposal](https://github.com/openshift/enhancements/blob/master/enhancements/monitoring/user-workload-monitoring.md) for more details. If enabled, only one stack can be deployed. The recommended way to get “platform” data is through Thanos Querier. Alternatively, the UWM stack can scrape the `/federate` endpoint expsed by the Platform Monitoring Prometheus to injest a subset of the available platform metrics.

Reviewer 20: ```suggestion
Please refer to the User Workload Monitoring [enhancement proposal](https://github.com/openshift/enhancements/blob/master/enhancements/monitoring/user-workload-monitoring.md) for more details. If enabled, only one stack can be deployed. The recommended way to get “platform” data is through Thanos Querier. Alternatively, the UWM stack can scrape the `/federate` endpoint exposed by the Platform Monitoring Prometheus to ingest a subset of the available platform metrics.
```

> Please refer to the User Workload Monitoring [enhancement proposal](https://github.com/openshift/enhancements/blob/master/enhancements/monitoring/user-workload-monitoring.md) for more details. If enabled, only one stack can be deployed. The recommended way to get “platform” data is through Thanos Querier. Alternatively, the UWM stack can scrape the `/federate` endpoint expsed by the Platform Monitoring Prometheus to injest a subset of the available platform metrics.

Author: ah we missed this, great point, let's clarify thanks for catching!

Moreover, Platform SRE expressed concerns that a shared Alertmanager configuration could lead to conflicts or misconfiguration that can impact alerting for every managed tenant running in the cluster.

Based on these discussions, we suggest that a dedicated Alertmanager instance is deployed with each `MonitoringStack` CR in the same namespace as the CR itself.
In next iteration we can focus on sharing AM across MonitoringStacks. We will take advantage of the recently added <code>[AlertmanagerConfig](https://github.com/prometheus-operator/prometheus-operator/blob/master/Documentation/api.md#alertmanagerconfig)</code> and allow MTO/MTSRE to create their own routing configuration in a centrally deployed Alertmanager instance.

### Integrating with Console

After discussions with console team we would like to avoid changing UX and add notion of data sources. Since OCS wants to integrate with OCP Console, we need to make sure Monitoring Stack addons can (opt-in) expose data to Platform Monitoring's Thanos Querier that is connected to Console.

> For this we added feature request for CMO recorded in [MON-1954](https://issues.redhat.com/browse/MON-1954). When implemented we will be able to deploy following architecture:

Reviewer 17: this might require more discussions at the implementation level but they can happen after this enhancement is accepted.

Author: 👍🏽

### Graduation Criteria

#### Dev Preview -> Tech Preview

* Monitoring Stack Operator can deploy dedicated Prometheus, Grafana DataSource, Thanos Query and Alertmanager.
* We are visible and installable from OperatorHub community catalog.

#### Tech Preview -> GA

* More options exposed in Monitoring Stack resource

> * We have onboard at lease few Addons on this feature.

Reviewer 17: "at lease" -> "at least"
